### PR TITLE
Fix different tooltips offset when hovering

### PR DESCRIPTION
**Pull request by Johann-S**
Fix different tooltips offset when hovering by resetting tip classes 

/CC : @mdo and @bardiharborow 

Close : #21146

### CODE REVIEW

**Comment by Johann-S** [comment]
The test failed on trying to stop the tunnel with SauceLabs
https://travis-ci.org/twbs/bootstrap/jobs/215178467#L431-L451

**Comment by Johann-S** [comment]
Squashed my commits and no more failed on Travis

File: js/src/tooltip.js [hunk 1]
```diff
@@ -438,6 +439,15 @@ const Tooltip = (($) => {
       return AttachmentMap[placement.toUpperCase()]
     }
 
+    _cleanTipClass() {
+      const $tip = $(this.getTipElement())
+      if (this.config.toggle === 'tooltip') {
+        $tip.attr('class', 'tooltip fade')
```
Comment (Johann-S) [inline js/src/tooltip.js]: good feedbacks @vanduynslagerp thank you 👍

File: js/src/tooltip.js [hunk 2]
```diff
@@ -438,6 +439,15 @@ const Tooltip = (($) => {
       return AttachmentMap[placement.toUpperCase()]
     }
 
+    _cleanTipClass() {
```
Comment (Johann-S) [inline js/src/tooltip.js]: Yes my bad I'll do that 😄

Comment (Johann-S) [inline js/src/tooltip.js]: Finally I found a generic solution but thank you @vanduynslagerp 👍

Comment (vanduynslagerp) [inline js/src/tooltip.js]: Using the `CLASS_PREFIX` is clever !
You can make this code shorter and more efficient thought:
```js
// On top of the file
const TETHER_PREFIX_REGEX = new RegExp(`(\S*)${CLASS_PREFIX}(\S*)`, 'g');

const $tip = $(this.getTipElement())
$tip.removeClass($tip.attr('class').match(TETHER_PREFIX_REGEX);
```
That makes the code more readable and  would avoid the multiple calls to Jquery to remove the classes one by one. It also avoid the loop and the assigning and reassigning of temporary variables multiple time.

Comment (Johann-S) [inline js/src/tooltip.js]: You're right nice regex thanks you @vanduynslagerp

**Comment by Johann-S** [comment]
I changed a bit your regex @vanduynslagerp but thank you it works very well 👍

**Comment by Johann-S** [comment]
This PR (https://github.com/twbs/bootstrap/pull/17094) tried to do the same thing but ``addTargetClasses`` only remove Tether class from the element which trigger the tooltip not the tooltip.

**Comment by vanduynslagerp** [comment]
> This PR (#17094) tried to do the same thing but addTargetClasses only remove Tether class from the element which trigger the tooltip not the tooltip.

I didn't check the Tether code, but is it possible that's it's actually a Tether bug ? It would be reasonable to think that when you activate the same element twice Tether should position it at the same position.
Or is the problem due to something Bootstrap does that prevent Tether to handle its own generated classes ?

**Comment by Johann-S** [comment]
Currently they are no way of cleaning the tooltip element in Tether and I didn't found any PR which target this issue on Tether repository. 
IMO it's due to Tether classes so until they don't fix this issue it's better to have this PR

**Comment by vanduynslagerp** [comment]
> Currently they are no way of cleaning the tooltip element in Tether and I didn't found any PR which target this issue on Tether repository.
IMO it's due to Tether classes so until they don't fix this issue it's better to have this PR

So in that case it might be better to open an issue with Tether and apply this PR if it's not fix fast enough? And we could remove this workaround once it's fixed ?

**Comment by Johann-S** [comment]
Finally after a quick search is possibly an issue in BS classes (but not sure about that) see : 
https://github.com/twbs/bootstrap/blob/v4-dev/scss/_tooltip.scss#L17-L59

FYI this issue exist since the merged of #17094 because before this PR BS4 cleaned Tether classes

**Comment by vanduynslagerp** [comment]
> Finally after a quick search is possibly an issue in BS classes (but not sure about that) see :
https://github.com/twbs/bootstrap/blob/v4-dev/scss/_tooltip.scss#L17-L59
FYI this issue exist since the merged of #17094 because before this PR BS4 cleaned Tether classes

Right ! Actually the code removed in #17094 was mentioning the Tether issue https://github.com/HubSpot/tether/issues/36, that is still opened....since March 2014...

**Comment by Johann-S** [comment]
Until they are a fix published by Tether I'll merge this PR.

Thank you for your help and feedbacks @vanduynslagerp 👍

**Comment by IdanCo** [comment]
@Johann-S kuddos for the fix, but even after the merge there is a strange behavior - when scrolling away and then back, the margin changes for the **top** and **left** tooltips. try scrolling down and up here -
http://jsbin.com/gopevakala/edit?html,output

I'm assuming it has something to do with the negative margins **top** and **left** tooltips has -
```sass
// scss/_tooltip.scss
  &.tooltip-top,
  &.bs-tether-element-attached-bottom {
    margin-top: -$tooltip-margin;
```
```sass
// scss/_tooltip.scss
  &.tooltip-left,
  &.bs-tether-element-attached-right {
    margin-left: -$tooltip-margin;
```

While right and bottom tooltips has positive margin -
```sass
// scss/_tooltip.scss
  &.tooltip-right,
  &.bs-tether-element-attached-left {
    margin-left: $tooltip-margin;
```
```sass
// scss/_tooltip.scss
  &.tooltip-bottom,
  &.bs-tether-element-attached-top {
    margin-top: $tooltip-margin;
```
Also, you can notice that the initial state lacks margins -
tooltips-margin

I tried looking for a quick win here by replacing the negative with positive, but the inconsistency continues. If you want i'll be happy to dig deeper.

**Comment by Johann-S** [comment]
Please @IdanCo open a new issue to expose this problem thank you 👍

**Pull request merged**